### PR TITLE
[RFC] Vim 7.4.{944,945,946,950,953,1032,1055}

### PR DESCRIPTION
**Pull request by watiko**
```
patch 7.4.944
Problem:    Writing tests for Vim script is hard.
Solution:   Add assertEqual(), assertFalse() and assertTrue() functions.  Add
            the v:errors variable.  Add the runtest script. Add a first new
            style test script.
```

https://github.com/vim/vim/commit/43345546ae63710441f066648b8485fb545b3801

```
patch 7.4.945
Problem:    New style testing is incomplete.
Solution:   Add the runtest script to the list of distributed files.
            Add the new functions to the function overview.
            Rename the functions to match Vim function style.
            Move undolevels testing into a new style test script.
```

https://github.com/vim/vim/commit/683fa185a4b4ed7595e5942901548b8239ed5cdb

```
patch 7.4.946
Problem:    Missing changes in source file.
Solution:   Include changes to the eval.c file.
```

https://github.com/vim/vim/commit/bbfbaf9741deebb9f1ed790885bd571c4cbce17a

```
patch 7.4.950
Problem:    v:errors is not initialized.
Solution:   Initialze it to an empty list. (Thinca)
```

https://github.com/vim/vim/commit/4649ded2877508fe343cbcf6f7e7fd277be0aab3

```
patch 7.4.953
Problem:    When a test script navigates to another buffer the .res file is
            created with the wrong name.
Solution:   Use the "testname" for the .res file. (Damien)
```

https://github.com/vim/vim/commit/de0ad40cb3c1bc691a754698ed16a5b6cdb4086b

```
patch 7.4.1032
Problem:    message from assert_false() does not look nice.
Solution:   Handle missing sourcing_name. Use right number of spaces. (Watiko)
            Don't use line number if it's zero.
```

https://github.com/vim/vim/commit/cbfe32953aea09d35d9ac7e5865c915b14e310c1

```
patch 7.4.1055
Problem:    Running "make newtests" in src/testdir has no output.
Solution:   List the messages file when a test fails. (Christian Brabandt)
            Update the list of tests.
```

https://github.com/vim/vim/commit/e7893a4088d6ea796bcab6195d232cb26c12c317

### CODE REVIEW

**Comment by watiko** [comment]
```
src/nvim/eval.c:378:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:378:  Missing space before }  [whitespace/braces] [5]
src/nvim/eval.c:379:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:379:  Missing space before }  [whitespace/braces] [5]
src/nvim/eval.c:380:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:380:  Missing space before }  [whitespace/braces] [5]
src/nvim/eval.c:7100:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:7100:  Missing space before }  [whitespace/braces] [5]
src/nvim/eval.c:7101:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:7101:  Missing space before }  [whitespace/braces] [5]
src/nvim/eval.c:7102:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:7102:  Missing space before }  [whitespace/braces] [5]
src/nvim/eval.c:7103:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:7103:  Missing space before }  [whitespace/braces] [5]
src/nvim/eval.c:7104:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:7104:  Missing space before }  [whitespace/braces] [5]
src/nvim/eval.c:7105:  Missing space after {  [whitespace/braces] [5]
src/nvim/eval.c:7105:  Missing space before }  [whitespace/braces] [5]
Total errors found: 18
CMake Error at cmake/RunLint.cmake:27 (message):
  Linting failed: 1.
```

I couldn't fix above linting errors.

```
...ild/neovim/neovim/test/functional/legacy/assert_spec.lua:7: Expected objects to be the same.
Passed in:
(table) {
 *[1] = '/private/tmp/lua_03z3JU line 1: 'equal assertion failed''
  [2] = '/private/tmp/lua_03z3JU line 2: 'true  assertion failed''
  [3] = '/private/tmp/lua_03z3JU line 3: 'false assertion failed''
  [4] = '/private/tmp/lua_UZ2HVB line 1: 'file two'' }
Expected:
(table) {
 *[1] = '/tmp/lua_03z3JU line 1: 'equal assertion failed''
  [2] = '/tmp/lua_03z3JU line 2: 'true  assertion failed''
  [3] = '/tmp/lua_03z3JU line 3: 'false assertion failed''
  [4] = '/tmp/lua_UZ2HVB line 1: 'file two'' }
```

It looks strange to me...

**Comment by justinmk** [comment]
If linting errors are too much, don't worry about it.

**Comment by justinmk** [comment]
Oh jeez, Bram added the test methods to the core? FFS.

**Comment by watiko** [comment]
Vim's new test will be written with assert_* functions, if neovim supports they, we can use legacy test cases to test porting patch of Vim at least. But I don't know we really needs to merge they.

**Comment by ZyX-I** [comment]
@watiko Why you can’t? Just modify the whole list of functions according to our style guide `{ "assert_true", 0, 1 },` (better in a separate commit).

@justinmk Plugins don’t need testing? Or neovim does not need to allow plugins to test themselves?

**Comment by watiko** [comment]
@ZyX-l I thought they shouldn't be changed (`couldn't` was not suitable word sorrry), but I will fix linting errors.

To get realpath fixes the `/private` prefix error but, I think, `lfs` is required. (https://github.com/pantsbuild/pants/pull/1994)

**Comment by ZyX-I** [comment]
@watiko Third reason why lint errors suppression was written the way it was is that it forces incremental code base improve of the old code: once you edit something near the error you also need to fix the error.

(Reasons were

1. Moving old code around inside a file (both by moving the old code directly and adding new code before the old thus shifting line numbers) does not make linter fail.
2. It is easy to suppress errors like this.
3. It makes it necessary to change the surrounding code, thus making codebase clearer without waiting for somebody to step in with a PR that just fixes linting errors (which are currently not accepted in any case).

)

**Comment by justinmk** [comment]
> Plugins don’t need testing? Or neovim does not need to allow plugins to test themselves?

Something like https://github.com/junegunn/vader.vim allows very nice plugin tests. It has some (few) limitations because internally it is using macros--but fixing those limiations in Vim would be a much better and more generally productive goal, _or_ providing some other mechanism for creating external test frameworks such as vader. Adding `assert_foo` and a new special-purpose global `v:errors` to the core, is cheesy. As if _these_ were the things holding back plugins from testing themselves? _This_ was the missing piece of the puzzle? Sigh.

But they were not guarded with a `FEAT_` flag so I guess we must include them. Yet more garbage in the Vim API.

**Comment by ZyX-I** [comment]
@justinmk *Any* external plugin may be ignored *just* because it is *external* and *unofficial*. Nobody doubts that there are plugins which allow testing, but to be compatible we must have `assert_*` family because I expect it to be more widely used in new plugins’ tests as presence of them in the API automatically makes them official way of testing plugins, and absense of `assert_*` will fail the tests.

**Comment by justinmk** [comment]
Yes, we must include the patch. I was just ranting.

File: test/functional/legacy/assert_spec.lua [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  local tmpname = os.tmpname()
+
+  if ffi.os == 'OSX' and string.match(tmpname, '^/tmp') then
+   tmpname = '/private'..tmpname
+  end
+
+  write_file(tmpname, code)
+  nvim.command('source '..tmpname)
+  os.remove(tmpname)
+  return tmpname
```
Comment (justinmk) [inline test/functional/legacy/assert_spec.lua]: Why not modify the `helpers.source()` function instead of redefining it here?

Comment (watiko) [inline test/functional/legacy/assert_spec.lua]: I thought this way is too ad-hoc to use for any functional test, but ok, I'll do that.

**Comment by watiko** [comment]
Vim's patch 7.4.1032 is merged into this PR.

**Comment by Shougo** [comment]
@watiko I think Vim 7.4.1058 and 7.4.1060 are needed pathces.

**Comment by watiko** [comment]
@Shougo Thanks for giving notice. I think this PR should not include Vim 7.4.1058(`alloc_fail()`).
`alloc_fail()` status looks WIP and We should make another PR.

Vim patch 7.4.10{55,60,71} are needed.

File: src/nvim/eval.c [hunk 2]
```diff
@@ -7981,6 +7986,110 @@ static void f_argv(typval_T *argvars, typval_T *rettv)
...
+
+  ga_init(gap, 1, 100);
+  if (sourcing_name != NULL) {
+    ga_concat(gap, sourcing_name);
+    if (sourcing_lnum > 0) {
+      ga_concat(gap, (char_u *)" ");
+    }
+  }
+  if (sourcing_lnum > 0) {
+    vim_snprintf(buf, ARRAY_SIZE(buf), "line %ld", (long)sourcing_lnum);
```
Comment (justinmk) [inline src/nvim/eval.c]: `%ld` should be a portable format specifier, see https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim#code-differences .

I will make this change.

File: test/functional/legacy/assert_spec.lua [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+describe('assert function:', function()
+
+  before_each(function()
+    clear()
+  end)
+
+  -- assert_equal({expected}, {actual}, [, {msg}])
+  describe('assert_equal', function()
+    it('should not change v:errors when expected is equal to actual', function()
```
Comment (justinmk) [inline test/functional/legacy/assert_spec.lua]: Really nice descriptions and structure in the tests you ported.

**Comment by justinmk** [comment]
@watiko Great PR, thanks!

File: src/nvim/testdir/Makefile [hunk 4]
```diff
@@ -147,3 +150,19 @@ test49.out: test49.vim
...
+
+# New style of tests uses Vim script with assert calls.  These are easier
+# to write and a lot easier to read and debug.
+# Limitation: Only works with the +eval feature.
+RUN_VIMTEST = VIMRUNTIME=$(SCRIPTSOURCE); export VIMRUNTIME; $(VALGRIND) $(VIMPROG) -u unix.vim -U NONE --noplugin
+
+newtests: newtestssilent
+	@/bin/sh -c "if test -f messages && grep -q 'FAILED' messages; then \
+	                 cat messages && cat test.log;                      \
+	             fi"                                                    \
```
Comment (watiko) [inline src/nvim/testdir/Makefile]: Oh...
s/fi"                                                    \\/fi"/

**Comment by Shougo** [comment]
@wakito OK.  Thanks.

**Pull request merged**